File: 3° semestre/Tecnicas-de-programacão-2/aula-24-08-23/Prototype/prototype01.py
import copy # n sei o q faz, se vira ae

class Pessoa:
    def __init__ (self, id, nome, idade):
        self.id = id
        self.nome = nome
        self.idade = idade

def clone(self):
    return copy.copy(self) # script do prof, o meu era diferente

# Classe gerenciamentoPessoa para gerenciar instancias de Pessoas:
class GerenciamentoPessoa:
    def __init__(self):
        self.pessoas = {}

    def adicionarPessoa(self, id, nome, idade):
        pessoa = Pessoa(id, nome, idade)
        self.pessoas[id] = pessoa

    def getPessoaById(id):
        pessoaOriginal = self.pessoas[id]
        if pessoaOriginal: return pessoaOriginal.clone()
        else: return None # n tem null, só none

# Criando 
gerencia = GerenciamentoPessoa()

gerencia.adicionarPessoa(1, 'junin', 21)
gerencia.adicionarPessoa(2, 'jobson', 40)
gerencia.adicionarPessoa(3, 'Edson', 30)

pessoaClone1 = gerencia.getPessoaById(1)
if pessoaClone1: pessoaClone1.nome = 'Junin Gameplays'

print('----------- Pessoa original -----------')
print(gerencia.getPessoaById(1).__dict__) # dict cria e grava um dicionario
print(gerencia.getPessoaById(2).__dict__)
print(gerencia.getPessoaById(3).__dict__)

print("----------- Pessoa clonada -----------")
print(pessoaClone1.__dict__)

File: 3° semestre/Tecnicas-de-programacão-2/aula-24-08-23/Prototype/prototype01.js
// classe pessoa - usada como referencia para ser clonada
class Pessoa {
    constructor(id, nome, idade) {
        this.id = id;
        this.nome = nome;
        this.idade = idade;
    }

    // metodo clone para realizar a copia do objeto
    clone() {
        return new Pessoa(this.id, this.nome, this.idade);
    }
}

// Classe GerenciamentoPessoa para gerenciar instancias de pessoas:
class GerenciamentoPessoa {
    constructor() {
        this.pessoas = {};
    }

    // adiciona uma nova pessoa ao dicionario de pessoas
    adicionarPessoa(id, nome, idade) {
        const pessoa = new Pessoa(id, nome, idade);
        this.pessoas[id] = pessoa;
    }

    // solicita uma pessoa pelo id e retorna uma copia dela:
    getPessoaById(id) {
        const pessoaOriginal = this.pessoas[id];
        if (pessoaOriginal) {
            return pessoaOriginal.clone();
        } else {
            return null;
        }
    }
}

// Criando uma instancia de gerenciamentoPessoa:
const gerencia = new GerenciamentoPessoa();

// Adicionando pessoas:
gerencia.adicionarPessoa(1, "junin", 21);
gerencia.adicionarPessoa(2, "jobson", 40);
gerencia.adicionarPessoa(3, "Edson", 30);

// Clonando pessoas e modificando informações:
const pessoaClone1 = gerencia.getPessoaById(1);
if (pessoaClone1) {
    pessoaClone1.nome = "Junin Gameplays";
}

// exibindo pessoas:
console.log("----------- Pesssoa original -----------");
console.log(gerencia.getPessoaById(1));
console.log(gerencia.getPessoaById(2));
console.log(gerencia.getPessoaById(3));

console.log("----------- Pesssoa clonada -----------");
console.log(pessoaClone1);